predicate/parser.go

```diff
@@ -0,0 +1,250 @@
+	p.n++
+}
+
+// curr returns the last read token.
+func (p *parser) curr() (tok influxql.Token, pos influxql.Pos, lit string) {
+	buf := &p.buf[(p.i-p.n+len(p.buf))%len(p.buf)]
+	return buf.tok, buf.pos, buf.lit
+}
+
+// scanIgnoreWhitespace scans the next non-whitespace token.
+func (p *parser) scanIgnoreWhitespace() (tok influxql.Token, pos influxql.Pos, lit string) {
+	tok, pos, lit = p.scan()
+	if tok == influxql.WS {
+		tok, pos, lit = p.scan()
+	}
+	return
+}
+
+// Parse the predicate statement.
+func Parse(sts string) (n Node, err error) {
+	if sts == "" {
+		return nil, nil
+	}
+	p := new(parser)
+	p.sc = influxql.NewScanner(strings.NewReader(sts))
+	return p.parseLogicalNode()
+}
+
+func (p *parser) parseLogicalNode() (Node, error) {
+	n := &LogicalNode{
+		Children: make([]Node, 0),
+	}
+	var currentOp LogicalOperator
+	for {
+		tok, pos, _ := p.scanIgnoreWhitespace()
+		switch tok {
+		case influxql.NUMBER:
+			fallthrough
+		case influxql.INTEGER:
+			fallthrough
+		case influxql.NAME:
+			fallthrough
+		case influxql.IDENT:
+			p.unscan()
+			tr, err := p.parseTagRuleNode()
+			if err != nil {
+				return *n, err
+			}
+			n.Children = append(n.Children, tr)
+		case influxql.AND:
+			if currentOp == 0 || currentOp == LogicalAnd {
+				currentOp = LogicalAnd
+				n.Operator = LogicalAnd
+			} else {
+				lastChild := n.Children[len(n.Children)-1]
+				var n1 Node
+				var err error
+				if tokNext := p.peekTok(); tokNext == influxql.LPAREN {
+					n1, err = p.parseLogicalNode()
+				} else {
+					n1, err = p.parseTagRuleNode()
+				}
+				if err != nil {
+					return *n, err
+				}
+				n.Children = append(n.Children[:len(n.Children)-1], LogicalNode{
+					Children: []Node{lastChild, n1},
+					Operator: LogicalAnd,
+				})
+			}
+		case influxql.OR:
+			if currentOp == 0 || currentOp == LogicalOr {
+				n.Operator = LogicalOr
+			} else {
+				n1, err := p.parseLogicalNode()
+				if err != nil {
+					return *n, err
+				}
+				n = &LogicalNode{
+					Children: []Node{*n, n1},
+					Operator: LogicalOr,
+				}
+			}
+			currentOp = LogicalOr
+		case influxql.LPAREN:
+			p.openParen++
+			currParen := p.openParen
+			n1, err := p.parseLogicalNode()
+			if err != nil {
+				return *n, err
+			}
+			if p.openParen != currParen-1 {
+				return *n, &influxdb.Error{
+					Code: influxdb.EInvalid,
+					Msg:  fmt.Sprintf("extra ( seen"),
+				}
+			}
+			n.Children = append(n.Children, n1)
+		case influxql.RPAREN:
+			p.openParen--
+			fallthrough
+		case influxql.EOF:
+			if p.openParen < 0 {
+				return *n, &influxdb.Error{
+					Code: influxdb.EInvalid,
+					Msg:  fmt.Sprintf("extra ) seen"),
+				}
+			}
+			if len(n.Children) == 1 {
+				return n.Children[0], nil
+			}
+			return *n, nil
+		default:
+			return *n, &influxdb.Error{
+				Code: influxdb.EInvalid,
+				Msg:  fmt.Sprintf("bad logical expression, at position %d", pos.Char),
+			}
+		}
+	}
+}
+
+func (p *parser) parseTagRuleNode() (TagRuleNode, error) {
```

Comment:
There is a lot of stuff here we will parse but not support. It seems to me like it would be better to fail at the parser stage.